Index: includes/menu.inc
===================================================================
--- includes/menu.inc	(revision 1861)
+++ includes/menu.inc	(working copy)
@@ -864,6 +864,10 @@
  *
  * @param $menu_name
  *   The named menu links to return
+ * @param $only_active_trail
+ *   Whether to only return the links in the active trail (TRUE) instead of all
+ *   links on every level of the menu link tree (FALSE). Defaults to FALSE.
+ *   Internally used for breadcrumbs only.
  * @return
  *   An array of menu links, in the order they should be rendered. The array
  *   is a list of associative arrays -- these have two keys, link and below.
@@ -871,7 +875,7 @@
  *   submenu below the link if there is one, and it is a subtree that has the
  *   same structure described for the top-level array.
  */
-function menu_tree_page_data($menu_name = 'navigation') {
+function menu_tree_page_data($menu_name = 'navigation', $only_active_trail = FALSE) {
   static $tree = array();
 
   // Load the menu item corresponding to the current page.
@@ -879,6 +883,18 @@
     // Generate a cache ID (cid) specific for this page.
     $cid = 'links:'. $menu_name .':page-cid:'. $item['href'] .':'. (int)$item['access'];
 
+    // If we are asked for the active trail only, and $menu_name has not been
+    // built and cached for this page yet, then this likely means that it
+    // won't be built anymore, as this function is invoked from
+    // template_process_page(). So in order to not build a giant menu tree that
+    // needs to be checked for access on all levels, we simply check whether we
+    // have the menu already in cache, or otherwise, build a minimum tree
+    // containing the breadcrumb/active trail only.
+    // @see menu_set_active_trail()
+    if (!isset($tree[$cid]) && $only_active_trail) {
+      $cid .= ':trail';
+    }
+
     if (!isset($tree[$cid])) {
       // If the static variable doesn't have the data, check {cache_menu}.
       $cache = cache_get($cid, 'cache_menu');
@@ -916,7 +932,7 @@
           $placeholders = implode(', ', array_fill(0, count($args), '%d'));
           $expanded = variable_get('menu_expanded', array());
           // Check whether the current menu has any links set to be expanded.
-          if (in_array($menu_name, $expanded)) {
+          if (!$only_active_trail && in_array($menu_name, $expanded)) {
             // Collect all the links set to be expanded, and then add all of
             // their children to the list as well.
             do {
@@ -943,7 +959,7 @@
         $data['tree'] = menu_tree_data(db_query("
           SELECT m.load_functions, m.to_arg_functions, m.access_callback, m.access_arguments, m.page_callback, m.page_arguments, m.title, m.title_callback, m.title_arguments, m.type, m.description, ml.*
           FROM {menu_links} ml LEFT JOIN {menu_router} m ON m.path = ml.router_path
-          WHERE ml.menu_name = '%s' AND ml.plid IN (". $placeholders .")
+          WHERE ml.menu_name = '%s' AND " . ($only_active_trail ? 'ml.mlid' : 'ml.plid') . " IN (". $placeholders .")
           ORDER BY p1 ASC, p2 ASC, p3 ASC, p4 ASC, p5 ASC, p6 ASC, p7 ASC, p8 ASC, p9 ASC", $args), $parents);
         $data['node_links'] = array();
         menu_tree_collect_node_links($data['tree'], $data['node_links']);
@@ -1019,7 +1035,6 @@
     }
   }
   _menu_tree_check_access($tree);
-  return;
 }
 
 /**
@@ -1510,7 +1525,7 @@
 }
 
 /**
- * Sets or gets the active trail (path to root menu root) of the current page.
+ * Sets or gets the active trail (path to menu tree root) of the current page.
  *
  * @param $new_trail
  *   Menu trail to set, or NULL to use previously-set or calculated trail. If
@@ -1563,7 +1578,10 @@
       }
     }
 
-    $tree = menu_tree_page_data(menu_get_active_menu_name());
+    // Pass TRUE for $only_active_trail to make menu_tree_page_data() build a
+    // stripped down menu tree containing the active trail only, in case the
+    // given menu has not been built in this request yet.
+    $tree = menu_tree_page_data(menu_get_active_menu_name(), TRUE);
     list($key, $curr) = each($tree);
 
     while ($curr) {
@@ -1612,7 +1630,7 @@
   }
 
   $item = menu_get_item();
-  if ($item && $item['access']) {
+  if (!empty($item['access'])) {
     $active_trail = menu_get_active_trail();
 
     foreach ($active_trail as $parent) {
